doc/manual.rst

```diff
@@ -3707,6 +3707,71 @@ location is derived from the second parameter (called
 ``varTy[T, 2]`` which is incompatible with ``varTy[T, 1]``.
 
 
+NRVO
+----
+
+The return value is represented inside the body of a routine as the special
+`result`:idx: variable. This allows for a mechanism much like C++'s
+"named return value optimization" (`NRVO`:idx:). NRVO means that the stores
+to ``result`` inside ``p`` directly affect the destination ``dest``
+in ``let/var dest = p(args)`` (definition of ``dest``) and also in ``dest = p(args)``
+(assignment to ``dest``). This is achieved by rewriting ``dest = p(args)``
+to ``p'(args, dest)`` where ``p'`` is a variation of ``p`` that returns ``void`` and
+receives a hidden mutable parameter representing ``result``.
+
+Informally:
+
+.. code-block:: nim
+  proc p(): BigT = ...
+
+  var x = p()
+  x = p()
+
+  # is roughly turned into:
+
+  proc p(result: var BigT) = ...
+
+  var x; p(x)
+  p(x)
+
+
+Let ``T``'s be ``p``'s return type. NRVO applies for ``T``
+if ``sizeof(T) >= N`` (where ``N`` is implementation dependent),
+in other words, it applies for "big" structures.
+
+If ``p`` can raise an exception, NRVO applies regardless. This can produce
```

Comment:
is this really going to be documented as a feature?

It's extremely surprising behaviour that breaks with the most basic foundations of what a function invocation is, sacrificing basic correctness in the name of a minor optimization - if this behaviour is wanted for whatever reason (in a case where the performance would matter), it's trivial to rewrite the function to take a `var` parameter to make the destructive behaviour clearly visible, but the bugs that will follow from this kind of approach will be hard to find in any larger codebase, and will worsen as more deterministic resource management is baked into the language - ie consider when a destructor has to roll back a partially updated instance due to a gotcha like this.

the compiler (in theory) can already remove most cases of the copying when it can prove that the result is not observable - ie it can apply to optimization to where the instance is "fresh" - `let x = xxx()` etc.

If anything, one could imagine a hint for when `NRVO` is _not_ applied could be useful - ie "missed optimization" warning.

Reply:
I think the place for discussion is here https://github.com/nim-lang/RFCs/issues/230